libraries/reliable_amqp_publisher/reliable_amqp_publisher.cpp

```diff
@@ -0,0 +1,236 @@
+   uint16_t onNegotiate(AMQP::TcpConnection* connection, uint16_t interval) override {
+      return 0;
+   }
+
+   void onError(AMQP::TcpConnection*, const char* message) override {
+      cbs->amqp_error(message);
+   }
+
+   reliable_amqp_publisher_callbacks* cbs;
+};
+
+struct reliable_amqp_publisher_impl final : reliable_amqp_publisher_callbacks {
+   reliable_amqp_publisher_impl(const std::string& url, const std::string& exchange, const boost::filesystem::path& unconfirmed_path) :
+     data_file_path(unconfirmed_path), exchange(exchange) {
+      std::string host = "localhost", user = "guest", pass = "guest", path;
+      uint16_t port = 5672;
+      fc::url parsed_url(url);
+      FC_ASSERT(parsed_url.proto() == "amqp", "Only amqp:// URIs are supported for AMQP addresses");
+      if(parsed_url.host() && parsed_url.host()->size())
+         host = *parsed_url.host();
+      if(parsed_url.user())
+         user = *parsed_url.user();
+      if(parsed_url.pass())
+         pass = *parsed_url.pass();
+      if(parsed_url.port())
+         port = *parsed_url.port();
+      if(parsed_url.path())
+         path = parsed_url.path()->string();
+
+      std::stringstream hostss;
+      hostss << "amqp://" << user << ":" << pass << "@" << host << ":" << port << "/" << path;
+      amqp_address = std::make_unique<AMQP::Address>(hostss.str());
+
+      if(boost::filesystem::exists(data_file_path)) {
+         try {
+            fc::datastream<fc::cfile> file;
+            file.set_file_path(data_file_path);
+            file.open("rb");
+            fc::raw::unpack(file, message_deque);
+         } FC_RETHROW_EXCEPTIONS(error, "Failed to load previously cached AMQP messages from ${f}", ("f", (fc::path)data_file_path));
+      }
+      boost::system::error_code ec;
+      boost::filesystem::remove(data_file_path, ec);
+
+      thread = std::thread([this]() {
+         try {
+            fc::set_os_thread_name("amqp");
+            retry_connection();
+            ctx.run();
+         }
+         catch(...) {
+            elog("Exception escaped from AMQP connection ${s} publishing to \"${e}\", this connection is permanently disabled now",
+                 ("s", (std::string)*amqp_address)("e", this->exchange));
+         }
+      });
+   }
+
+   ~reliable_amqp_publisher_impl() {
+      stopping = true;
+
+      //drain any remaining items on queue
+      std::promise<void> shutdown_promise;
+      ctx.post([&]() {
+         shutdown_promise.set_value();
+      });
+      shutdown_promise.get_future().wait();
+
+      ctx.stop();
+      thread.join();
+
+      //if a message is in flight, let it try to run to completion
+      if(in_flight) {
+         auto until = std::chrono::steady_clock::now() + std::chrono::seconds(5);
+         ctx.restart();
+         while(in_flight && ctx.run_one_until(until));
+      }
+
+      if(message_deque.size()) {
+         try {
+            boost::filesystem::create_directories(data_file_path.parent_path());
```

Comment:
It needs to capture exception or handle error code for create_directories()

Reply:
It's inside an FC_LOG_AND_DROP(), not really sure what else can be done here on failure since we're shutting down.